pkg/apis/flowcontrol/types.go

```diff
@@ -88,10 +93,11 @@ type FlowSchemaSpec struct {
 	PriorityLevelConfiguration PriorityLevelConfigurationReference
 	// `matchingPrecedence` is used to choose among the FlowSchemas that match a given request. The chosen
 	// FlowSchema is among those with the numerically lowest (which we take to be logically highest)
-	// MatchingPrecedence.  Each MatchingPrecedence value must be non-negative.
-	// Note that if the precedence is not specified or zero, it will be set to 1000 as default.
+	// MatchingPrecedence.  Each MatchingPrecedence value must be ranged in [0,10000]. For two FlowSchemas
+	// with conflicting MatchingPrecedence, the one created earlier is prior.
```

Comment:
@deads2k had a good point that this is not stable if we add the same rules to multiple clusters.

Here's another idea: what if, if there are N matching and tied FSCs, we choose one randomly with probability 1/N? @MikeSpreitzer @yue9944882 

My thinking is that alphabetical disambiguation doesn't really make any logical sense as a priority mechanism. The coin-flip as a tie-break has plenty of historical precedent and also behaves the same on all clusters (at least in aggregate). We need to help admins know that their configuration has this condition, but we'd want to do that anyway.

(Actually I can see folks using this on purpose to e.g. test a new F.S. before deleting a previous one)

Reply:
Everything else about this PR LGTM.

Reply:
The flow schemas should be in a clearly defined order, and that's what the MatchingPrecedence values are for.  There should not be two FlowSchema objects with the same MatchingPrecedence value, but our implementation has to do _something_ in this case.  My position is that we should tell users to avoid this case and make no promise about what happens in this case, and I do not care what our implementation actually does.  Frankly, the more user-hostile the better, because this is a case that the users should avoid.

Reply:
> The flow schemas should be in a clearly defined order, and that's what the MatchingPrecedence values are for.

+1. the behavior of flow-schema matching should be certain, no matter if the logic makes sense or not. what happening to one request through the shuffle-sharding + fair-queuing is already tough to observe, it can be worse if we randomlize the order..